jupyter/d2l-java/chapter_multilayer-perceptrons/underfit-overfit.ipynb

```diff
@@ -0,0 +1,981 @@
+    "\n",
+    "Arrays.fill(lossLabel, 0, testLoss.length, \"test loss\");\n",
+    "Arrays.fill(lossLabel, testLoss.length, trainLoss.length + testLoss.length, \"train loss\");\n",
+    "\n",
+    "Table data = Table.create(\"Data\").addColumns(\n",
+    "      FloatColumn.create(\"epochCount\", ArrayUtils.addAll(epochCount, epochCount)), \n",
+    "      FloatColumn.create(\"loss\", ArrayUtils.addAll(testLoss, trainLoss)),\n",
+    "      StringColumn.create(\"lossLabel\", lossLabel)\n",
+    ");\n",
+    "\n",
+    "render(LinePlot.create(\"\", data, \"epochCount\", \"loss\", \"lossLabel\"),\"text/html\");"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Insufficient Training (Overfitting)\n",
+    "\n",
+    "Now let us try to train the model\n",
+    "using a polynomial of too high degree.\n",
+    "Here, there is insufficient data to learn that\n",
+    "the higher-degree coefficients should have values close to zero.\n",
+    "As a result, our overly-complex model\n",
+    "is far too susceptible to being influenced\n",
+    "by noise in the training data.\n",
+    "Of course, our training error will now be low\n",
+    "(even lower than if we had the right model!)\n",
+    "but our test error will be high.\n",
+    "\n",
+    "Try out different model complexities (`nDegree`)\n",
+    "and training set sizes (`nSubset`)\n",
+    "to gain some intuition of what is happening."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "8"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "int nSubset = 100;\n",
+    "int nDegree = 20;\n",
+    "train(polyFeatures.get(\"1:\" + nSubset + \", 0:\"+ nDegree), \n",
+    "                      polyFeatures.get(nTrain + \": , 0:\"+ nDegree), \n",
+    "                      labels.get(\"1:\" + nSubset), \n",
+    "                      labels.get(nTrain + \":\"), nDegree);"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "System.out.println(\"weight: \" + weight);"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "String[] lossLabel = new String[trainLoss.length + testLoss.length];\n",
+    "\n",
+    "Arrays.fill(lossLabel, 0, testLoss.length, \"test loss\");\n",
+    "Arrays.fill(lossLabel, testLoss.length, trainLoss.length + testLoss.length, \"train loss\");\n",
+    "\n",
+    "Table data = Table.create(\"Data\").addColumns(\n",
+    "      FloatColumn.create(\"epochCount\", ArrayUtils.addAll(epochCount, epochCount)), \n",
+    "      FloatColumn.create(\"loss\", ArrayUtils.addAll(testLoss, trainLoss)),\n",
+    "      StringColumn.create(\"lossLabel\", lossLabel)\n",
+    ");\n",
+    "\n",
+    "render(LinePlot.create(\"\", data, \"epochCount\", \"loss\", \"lossLabel\"),\"text/html\");"
```

Comment:
Doesn't seemed to be underfitting. Can you do log scale on y?

Reply:
This actually varies with the values you get for the labels, features initialized above. Sometimes, the book version doesn't seem the same as when it run it on collab. Verified it there by running it multiple times.

Reply:
The result does showing underfitting, but the scale makes it looks the same as the normal one